ot/mpz-ot-core/src/ferret/spcot/receiver.rs

```diff
@@ -0,0 +1,293 @@
+use itybity::ToBits;
+use mpz_core::{aes::FIXED_KEY_AES, ggm_tree::GgmTree, hash::Hash, prg::Prg, Block};
+use rand_core::SeedableRng;
+
+use super::msgs::{
+    CheckFromReceiver, CheckFromSender, CotMsgForReceiver, ExtendFromSender, MaskBits,
+};
+
+/// SPCOT receiver.
+#[derive(Debug, Default)]
+pub struct Receiver<T: state::State = state::Initialized> {
+    pub(crate) state: T,
+}
+
+impl Receiver {
+    /// Creates a new Receiver.
+    pub fn new() -> Self {
+        Receiver {
+            state: state::Initialized::default(),
+        }
+    }
+
+    /// Completes the setup phase of the protocol.
+    ///
+    /// See step 1 in Figure 6.
+    ///
+    /// # Arguments
+    ///
+    /// * `seed` - The random seed to generate PRG.
+    pub fn setup(self, seed: Block) -> Receiver<state::Extension> {
+        Receiver {
+            state: state::Extension {
+                ws: Vec::default(),
+                chis: Vec::default(),
+                cot_counter: 0,
+                exec_counter: 0,
+                prg: Prg::from_seed(seed),
+            },
+        }
+    }
+}
+
+impl Receiver<state::Extension> {
+    /// Performs the mask bit step in extension.
+    ///
+    /// See step 4 in Figure 6.
+    ///
+    /// # Arguments
+    ///
+    /// * `h` - The depth of the GGM tree.
+    /// * `alpha` - The chosen position.
+    /// * `extend` - The message from COT ideal functionality for the receiver. Only the random bits are used.
+    pub fn extend_mask_bits(
+        &mut self,
+        h: usize,
+        alpha: usize,
+        extend: CotMsgForReceiver,
+    ) -> Result<MaskBits, ReceiverError> {
+        if alpha > (1 << h) {
+            return Err(ReceiverError::InvalidInput(
+                "the input pos should be no more than 2^h".to_string(),
+            ));
+        }
+
+        let CotMsgForReceiver { rs, ts: _ } = extend;
+
+        if rs.len() != h {
+            return Err(ReceiverError::InvalidLength(
+                "the length of b should be h".to_string(),
+            ));
+        }
+
+        // Step 4 in Figure 6
+        let mut alpha_vec = alpha.to_msb0_vec();
+        alpha_vec.drain(0..alpha_vec.len() - h);
+
+        let bs: Vec<bool> = alpha_vec
+            .iter()
+            // Computes alpha_i XOR r_i XOR 1.
+            .zip(rs.iter())
+            .map(|(alpha, r)| alpha == r)
+            .collect();
+
+        Ok(MaskBits { bs })
+    }
+
+    /// Performs the GGM reconstruction step in extension.
+    ///
+    /// See step 5 in Figure 6.
+    ///
+    /// # Arguments
+    ///
+    /// * `h` - The depth of the GGM tree.
+    /// * `alpha` - The chosen position.
+    /// * `extendfc` - The message from COT ideal functionality for the receiver. Only the chosen blocks are used.
+    /// * `extendfr` - The message sent from the sender.
+    pub fn extend(
+        &mut self,
+        h: usize,
+        alpha: usize,
+        extendfc: CotMsgForReceiver,
```

Comment:
Instead of depending on a concrete message type here, we can instead just depend on the correlations directly.

```suggestion
        ts: &[Block]
```